### PR TITLE
feat: network upgrade to async

### PR DESCRIPTION
**Pull request by driftluo**
network upgrade to async:
- [x] wait `molecule` release
- [x] wait `p2p` release

### CODE REVIEW

**Comment by driftluo** [comment]
These two have nothing to do with the p2p crate and should not be marked

**Comment by yangby-cryptape** [comment]
> These two have nothing to do with the p2p crate and should not be marked

My mistakes! I added two incorrect comments.

Exclude the two generated files and `Cargo.lock`, that PR has 40 files.
I rebased three times, it's too difficult to check all files after each rebase.

File: network/src/protocols/discovery.rs [hunk 1]
```diff
@@ -55,17 +61,17 @@ impl ServiceProtocol for DiscoveryProtocol {
...
-                    .for_each(|()| Ok(()))
-                    .map_err(|err| {
-                        warn!("discovery stream error: {:?}", err);
-                    })
-                    .then(|_| {
-                        debug!("End of discovery");
-                        Ok(())
-                    })
+                async move {
+                    loop {
```
Comment (doitian) [inline network/src/protocols/discovery.rs]: Any reason that we have to replace all the `for_each`s with an async loop?

Comment (driftluo) [reply network/src/protocols/discovery.rs]: This is a question of writing style, the two are equivalent

**Comment by yangby-cryptape** [comment]
bors r=yangby-cryptape,doitian

**Comment by driftluo** [comment]
bors retry

**Comment by driftluo** [comment]
rebase and fix test

**Comment by yangby-cryptape** [comment]
bors r=doitian,yangby-cryptape

**Comment by driftluo** [comment]
bors retry

**Comment by driftluo** [comment]
bors try

**Comment by driftluo** [comment]
bors try

**Comment by doitian** [comment]
bors r=doitian,yangby-cryptape

**Comment by doitian** [comment]
bors r=quake,yangby-cryptape

**Comment by doitian** [comment]
@driftluo please resolve conflicts

**Comment by driftluo** [comment]
> @driftluo please resolve conflicts

p2p will release the next version soon, do we need to merge and upgrade?

**Pull request merged**